### DISCUSSION TITLE
Enable PATCH Endpoint to Accept Arrays for Bulk Updates (Update Many Items with Unique Values Using Unique Key, Composite Key, or Query)

### DISCUSSION THREAD

**Discussion by JayShoe** [directus/directus#24083] [Feature Requests]
Currently, the POST endpoint supports bulk updates by accepting an array of objects, allowing many items to be created in a single request. In contrast, the PATCH endpoint does not support this functionality. It requires either multiple PATCH calls—one per item—or a single call that updates multiple items with the same values. This limitation drastically increases the number of API calls required for unique updates, impacting performance, scalability, and efficiency. Enhancing PATCH to accept arrays, like POST, would streamline bulk updates and reduce system overhead.

---

### **Basic Example**

**Endpoint URL:**
```http
PATCH /items/:collection
```

**Request Payload (Array of Updates):**
```json
[
    {
        "query": {
            "filter": {
                "id": {
                    "_eq": "999999"
                }
            },
            "limit": 1
        },
        "data": {
            "name": "Updated Name",
            "description": "Updated Description"
        }
    },
    {
        "query": {
            "filter": {
                "id": {
                    "_eq": "999998"
                }
            },
            "limit": 1
        },
        "data": {
            "name": "Updated Name 2",
            "description": "Updated Description 2"
        }
    }
]
```

**SQL Equivalent:**
```sql
UPDATE items
SET name = 'Updated Name', description = 'Updated Description'
WHERE id = '999999';

UPDATE items
SET name = 'Updated Name 2', description = 'Updated Description 2'
WHERE id = '999998';
```

- The API processes each object in the array independently, applying the `query` filters and `data` updates for each specified record.

---

### **Motivation**

Currently, the PATCH endpoint allows updating a single object using a query filter, but bulk updates require either multiple API calls or a single object with generalized updates that affect many items. This limitation increases complexity and reduces efficiency when dealing with unique updates for multiple items.

Supporting arrays in the PATCH payload would:

1. **Reduce API Overhead**:
   - Avoid the need for multiple PATCH calls when updating multiple unique records, significantly reducing the number of network requests.

2. **Simplify Bulk Operations**:
   - Streamline workflows where each record has unique updates.

3. **Improve Performance**:
   - Minimize latency and server load by consolidating multiple updates into one request.

---

### **Detailed Design**

1. **Behavior**:
   - The PATCH endpoint accepts an array of objects where each object specifies a `query` for filtering and the `data` to update.
   - The PATCH endpoint continues to accept a single object, preserving existing behavior for:
     - `PATCH /items/resource/:id` (direct update by primary key).
     - Single "patch-by-query" operations that apply one update to multiple records.
   - The PATCH endpoint will now support both:
     - A single object (one query, patch many).
     - An array of objects (multiple queries with respective updates for each).

2. **Validation**:
   - Each object in the array must include:
     - A `query` field with valid filters.
     - A `data` field specifying the update values.
   - If any object in the array is invalid, the API returns an error with details for the problematic entry while continuing to process the valid ones.

3. **Batch Execution**:
   - The server processes each object in the array independently, applying updates sequentially or in parallel, depending on the implementation.

4. **Response**:
   - Returns a summary of the update results, including:
     - The number of successful updates.
     - Details for any errors encountered during processing.

---

#### **Must Have**:
- Support for arrays in the PATCH payload.
- Validation for each object in the array, ensuring `query` and `data` fields are present and well-formed.
- Comprehensive error handling for invalid entries.

#### **Should Have**:
- Support for processing updates in parallel to improve performance for large arrays.

#### **Could Have**:
- Detailed response indicating the success or failure of each object update.

#### **Won't Have**:
- Changes to existing functionality for single-object updates.

---

### **Drawbacks**

1. **Implementation Complexity**:
   - Adding array processing increases the complexity of the PATCH endpoint logic.
2. **Error Handling**:
   - Detailed error responses for partial failures would complicate client-side handling, if implemented. Otherwise, a single error would suffice. 

---

### **Alternatives**

1. **Multiple PATCH Requests**:
   - Current functionality requires a PATCH call for each unique record.
   - **Impact**: For 100 records, this requires 100 API calls, increasing network overhead.

2. **Generalized Updates**:
   - A single PATCH call can update multiple records with shared criteria and data.
   - **Impact**: Cannot apply unique updates to multiple records, limiting flexibility.

---

### **Adoption Strategy**

1. **Backward Compatibility**:
   - Existing single-object PATCH requests remain unaffected.
2. **Incremental Adoption**:
   - Developers can begin using arrays as needed for bulk updates without altering current workflows.
3. **Documentation**:
   - Provide clear examples of both single-object and array-based PATCH requests to facilitate adoption.

---

### **Unresolved Questions**

1. **Concurrency**:
   - Should the updates be processed sequentially or in parallel?
2. **Batch Size**:
   - Is there a limit to the number of objects in the array for a single request?

**Comment by rijkvanzanten** [comment]
This reminds me of updating-by-query! 

`PATCH /items/:collection`

```json
{
  "query": {
    "filter": {
      "my_unique_field": {
        "_eq": "value"
      }
    },
    "limit": 1
  },
  "data": {
    "name": "Updated Name",
    "description": "Updated Description"
  }
}
```

**Reply by rijkvanzanten** [reply]
The same works for delete 🙂 It seems... super missing.. from the docs however 😨 cc @phazonoverload 🚨

**Reply by JayShoe** [reply]
That's amazing. I'll have to give it a try. 

Thanks @rijkvanzanten

**Reply by rijkvanzanten** [reply]
No prob! Little surprised it ~~somehow slipped through the cracks in our reference~~, but that'll be a quick fix 🙂 ref https://github.com/directus/directus/issues/24092

Update; it's actually in there, but only mentioned in passing "set either `keys` _or `query`_" very easy to miss; patch incoming 🚀

**Comment by JayShoe** [comment]
@rijkvanzanten can you provide an example for doing this with multiple items in an array? I tried the following.

`PATCH /items/item`

```
[
{
  "query": {
    "filter": {
      "id": {
        "_eq": "999999"
      }
    },
    "limit": 1
  },
  "data": {
    "name": "Updated Name",
    "description": "Updated Description"
  }
},
{
  "query": {
    "filter": {
      "id": {
        "_eq": "999998"
      }
    },
    "limit": 1
  },
  "data": {
    "name": "Updated Name 2",
    "description": "Updated Description 2"
  }
}
]
```

The error:
```
{
    "errors": [
        {
            "message": "Invalid payload. Item in update misses primary key.",
            "extensions": {
                "reason": "Item in update misses primary key",
                "code": "INVALID_PAYLOAD"
            }
        }
    ]
}

```

The error led me to believe this feature doesn't exist.

**Reply by rijkvanzanten** [reply]
Yeah that's not possible in this format. This update flow updates multiple items as selected by the query to the same data.

**Reply by JayShoe** [reply]
I guess my original concept had some merit. I will edit and update the request with this new info.

**Comment by JayShoe** [comment]
@rijkvanzanten

Thank you for your help in clarifying the current limitations with PATCH. I’ve updated my original request to reflect these limitations. I appreciate your guidance in helping me better understand the current status of this feature.

**Reply by rijkvanzanten** [reply]
Ty! 🙌🏻 I'm curious if there's a world where there's 'just' one endpoint to do multiple calls, almost like a gRPC endpoint. Technically speaking GraphQL already handles multiple operations as part of one network call, but it's not really that common to see in REST APIs.

**Comment by rijkvanzanten** [comment]
For what it's worth:

> This limitation drastically increases the number of API calls required for unique updates, impacting performance, scalability, and efficiency.

More network calls doesn't necessarily mean worse performance / scalability. If anything, in a load-balanced environment it can _improve_ scalability, as each call is lighter to execute and can be parallelized across each load balanced process handling the requests. If all requests are being handled by the same receiving server you end up putting more load on the single process handling that request 🙂 As a rule of thumb, I'd actually err on the side of doing more smaller API calls than fewer larger calls, especially if your load balancer/cdn stack supports http/2

**Reply by JayShoe** [reply]
That's an interesting point, @rijkvanzanten, and definitely worth considering.

Playing devil's advocate here—does the receiving API know how to handle chunks effectively?

> Batch Execution: The server processes each object in the array independently, applying updates sequentially or in parallel, depending on the implementation.

If that's the case, the receiving API should be able to throttle the batch as necessary, right? I’d imagine the handling of 100 updates in a single chunk would be fairly comparable to multiple smaller interactions, like ping/pong exchanges.

Two key points I didn’t explicitly mention in my original post:

1. ETL with Flows: I’m using flows to perform ETL operations, reformatting the data on its way to the new schema. Without a batch processing option, I need to implement a Trigger Flow followed by another flow to handle each individual PATCH. This significantly increases the complexity of these calls. Especially when you consider the complexity of managing a separate flow for one task.


2. Pressure Errors: I've noticed that without introducing a 10-50ms delay per call, I frequently encounter "under pressure" errors. This introduces a significant delay, impacting overall speed. I noticed on my logs each update taking about 9.5ms yet 100 posts complete in 45ms. The server is more responsive (docker desktop) when I do one call with 100 posts vs 100 calls. 


That's my experience so far. A separate note is that the upsert functionality, combined with this feature, would mean the PATCH can work as a POST too, so then we go from at least 3-4 complicated flow steps (find dup, check dup, separate post/patch items, post, patch) to literally 1 call to post/update 100 items. 

https://github.com/directus/directus/discussions/5706

**Reply by rijkvanzanten** [reply]
> If that's the case, the receiving API should be able to throttle the batch as necessary, right? I’d imagine the handling of 100 updates in a single chunk would be fairly comparable to multiple smaller interactions, like ping/pong exchanges.

A receiving API can indeed chunk the input batch and then throttle the execution to avoid issues, but that would still mean you're queuing those chunks on a single process. That in turn means you lose any ability to parallelize as node js is single threaded 🤔 Doing 100 updates as 100 requests means those requests can be spread across many worker processing handling those updates in parallel, whereas doing those 100 updates in a single request mean that those updates are handled by the same worker process handling the request, meaning it'll have to queue those sequentially. 

Thanks for the additional context 🙂 

The under pressure stuff is there because NodeJS doesn't really handle cycle pressure related issues natively all that well. It was an attempt at stopping the node event queue from overloading, but it kicks in a little too over-eager. It also has been causing some issues before on hosting providers where the CPU cycles are allocated virtually. Those hosts will often make more CPU power available if and when the process uses it, which in turn means the pressure backoff can kick in before the CPU power is increased. It's worth fiddling with the numbers and/or turning it off altogether. Depending on your hosting environment, it might improve things 🙂 I've been considering disabling it by default or even removing it altogether. Tricky one.. It helps a lot in some specific cases, but can hurt in others 😵

**Reply by JayShoe** [reply]
I updated my comment to refer to the flow steps not the call amount. The call amount doesn't matter to me. The fact I must create 5 steps (referenced above) instead of 1 step is why this interests me. 

Also, thanks for letting me know that I can disable the under pressure errors. I will definitely do that in my local environment. And also understanding what you mean about dividing the threads is making me believe that I could make four calls of 100 at the same time. So thanks for that note too. 👍🏻👍🏻👍🏻 Maybe I can optimize this a bit.
